### PR TITLE
feat(replica): add command to get rebuild estimation

### PR DESCRIPTION
**Pull request by utkarshmani1997**
This commit adds code to get rebuild estimation through jivactl command.
A new command syncinfo is added to jivactl.
Exp:
```
infinity:jiva(sync-progress)$ ./jiva syncinfo
Sync process not yet started
infinity:jiva(sync-progress)$ ./jiva syncinfo
DegradedReplica:  192.168.43.203, WOSnapshotsTotalSize:  29 MiB
HealthyReplica:   127.0.0.1,      RWSnapshotsTotalSize:  39 MiB
============================================================================
Snapshot                                 Status     DegradedSize HealthySize
ab5450c3-5fe3-43da-9fcb-1599eebbe732.img Completed  0 B          0 B
3c9ac3e4-ab4d-4bec-aa1e-4979384322e6.img InProgress 29 MiB       39 MiB
4793eabf-b42b-4b81-bca5-0faf3a2cbdf2.img Pending    NA           0 B
deba8ac7-4273-4ecc-996d-915024bce2f5.img Pending    0 B          0 B
infinity:jiva(sync-progress)$ ./jiva syncinfo
No replicas are undergoing sync process
```
Note: This command will not work when sync in progress for clone replica. It will be addressed in other PR.
Signed-off-by: Utkarsh Mani Tripathi <utkarsh.tripathi@mayadata.io>

### CODE REVIEW

File: app/replica.go [hunk 1]
```diff
@@ -157,6 +157,7 @@ func CloneReplica(s *replica.Server, address string, cloneIP string, snapName st
 	var err error
```
Comment (vishnuitta) [inline app/replica.go:157]: is this CloneReplica only for Clone related volumes?

Comment (utkarshmani1997) [reply app/replica.go:157]: yes

**Comment by utkarshmani1997** [comment]
@vishnuitta since rebuilding is skipped in healthy replica and also in WO replica in case revision count and chain is same as of healthy replica. So in this scenario i'm returning error. What's your opinion?
in other words what would be better approach?

**Comment by vishnuitta** [comment]
> @vishnuitta since rebuilding is skipped in healthy replica and also in WO replica in case revision count and chain is same as of healthy replica. So in this scenario i'm returning error. What's your opinion?
> in other words what would be better approach?

If sync is in progress (irrespective of whether chain is same or revisioncounter is same), put the details as mentioned in the table.
If sync is not in progress, put the message that 'no sync in progress' (not exact message).

File: app/syncinfo.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	replicaClient "github.com/openebs/jiva/replica/client"
+	"github.com/openebs/jiva/types"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+func SyncInfoCmd() cli.Command {
+	return cli.Command{
+		Name:      "syncinfo",
+		ShortName: "ls",
```
Comment (vishnuitta) [inline app/syncinfo.go:18]: this needs change

File: app/syncinfo.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		Action: func(c *cli.Context) {
+			if err := getSyncInfo(c); err != nil {
+				logrus.Fatalf("Error running ls command: %v", err)
```
Comment (vishnuitta) [inline app/syncinfo.go:21]: here also

File: app/syncinfo.go (hunk 2 above)
Comment (utkarshmani1997) [reply app/syncinfo.go:18]: good catch

File: replica/client/client.go [hunk 4]
```diff
@@ -463,6 +470,14 @@ func (c *ReplicaClient) get(url string, obj interface{}) error {
 		return err
 	}
 	defer resp.Body.Close()
+	if resp.StatusCode >= 300 {
```
Comment (vishnuitta) [inline replica/client/client.go:473]: do we still need these changes?

File: app/syncinfo.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	fmt.Fprintf(tw, "%v\t%v,\t%v\t%v\n", "DegradedReplica: ", strings.
+		TrimSuffix(strings.TrimPrefix(info.WOReplica, "http://"), ":9502/v1"),
+		"WOSnapshotsTotalSize: ", info.WOSnapshotsTotalSize)
+	fmt.Fprintf(tw, "%v\t%v,\t%v\t%v\n", "HealthyReplica: ", strings.
+		TrimSuffix(strings.TrimPrefix(info.RWReplica, "http://"), ":9502/v1"),
+		"RWSnapshotsTotalSize: ", info.RWSnapshotsTotalSize)
+	fmt.Fprintf(tw, "%s\n", "============================================================================")
+	fmt.Fprintf(tw, format, "Snapshot", "Status", "DegradedSize", "HealthySize")
```
Comment (vishnuitta) [inline app/syncinfo.go:63]: DegradedSnapSize, HealthySnapSize

File: app/syncinfo.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		fmt.Println("No replicas are undergoing sync process")
+		return nil
+	} else if info.Snapshots == nil {
+		fmt.Println("Sync process not yet started")
+		return nil
+	}
+
+	fmt.Fprintf(tw, "%v\t%v,\t%v\t%v\n", "DegradedReplica: ", strings.
+		TrimSuffix(strings.TrimPrefix(info.WOReplica, "http://"), ":9502/v1"),
+		"WOSnapshotsTotalSize: ", info.WOSnapshotsTotalSize)
```
Comment (vishnuitta) [inline app/syncinfo.go:58]: while printing, lets make it consistent.. WOSnapshotsTotalSize to DegradedSnapshotsTotalSize and also the next one..
(or) every where change to RW and WO from Healthy and Degraded

File: replica/client/client.go (hunk 4 above)
Comment (utkarshmani1997) [reply replica/client/client.go:473]: yes, good to have

File: app/syncinfo.go (hunk 5 above)
Comment (utkarshmani1997) [reply app/syncinfo.go:63]: roger that

File: app/syncinfo.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+func getSyncInfo(c *cli.Context) error {
+	controllerClient := getCli(c)
+
+	reps, err := controllerClient.ListReplicas()
+	if err != nil {
+		return err
+	}
+
+	var info *types.SyncInfo
+	format := "%v\t%v\t%v\t%v\n"
```
Comment (vishnuitta) [inline app/syncinfo.go:36]: You may need to add one more '\t' for snapshot column as snapshots can be bigger

File: app/syncinfo.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	tw := tabwriter.NewWriter(os.Stdout, 0, 20, 1, ' ', 0)
```
Comment (vishnuitta) [inline app/syncinfo.go:37]: why to use tabwriter? can't fmt printf helps?

File: app/syncinfo.go (hunk 7 above)
Comment (utkarshmani1997) [reply app/syncinfo.go:36]: tabwriter package will automatically adjust padding as set.

File: app/syncinfo.go (hunk 8 above)
Comment (utkarshmani1997) [reply app/syncinfo.go:37]: tabwriter package is useful in pretty printing the output. It takes care of padding and all. in fmt.printf we need to adjust for the small names and their sizes.

File: app/syncinfo.go (hunk 6 above)
Comment (utkarshmani1997) [reply app/syncinfo.go:58]: Sounds good

File: replica/rest/model.go [hunk 9]
```diff
@@ -293,6 +298,9 @@ func NewSchema() *client.Schemas {
 	schemas.AddType("replicaCounter", ReplicaCounter{})
 	schemas.AddType("replacediskInput", ReplaceDiskInput{})
 
+	rebuild := schemas.AddType("rebuildinfo", RebuildInfoOutput{})
```
Comment (vishnuitta) [inline replica/rest/model.go:301]: @payes can you review this REST related code?

File: app/syncinfo.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	for _, r := range reps {
+		if r.Mode == "ERR" || r.Mode == "RW" {
```
Comment (vishnuitta) [inline app/syncinfo.go:39]: just checking for r.Mode for !WO better

File: app/syncinfo.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+			continue
+		}
+		info, err = getRebuildInfo(r.Address)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (vishnuitta) [inline app/syncinfo.go:46]: break from here

File: app/syncinfo.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+	if info == nil {
+		fmt.Println("No replicas are undergoing sync process")
+		return nil
+	} else if info.Snapshots == nil {
+		fmt.Println("Sync process not yet started")
```
Comment (vishnuitta) [inline app/syncinfo.go:52]: lets print the IP which is in r.Address

File: sync/rebuild/progress.go [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/openebs/jiva/replica"
+	"github.com/openebs/jiva/types"
+	"github.com/openebs/jiva/util"
+)
+
+var (
+	Info *types.SyncInfo
+)
+
+func AddSyncInfo(syncInfo *types.SyncInfo) {
```
Comment (vishnuitta) [inline sync/rebuild/progress.go:16]: call it as SetSyncInfo rather than AddSyncInfo

File: sync/sync.go [hunk 14]
```diff
@@ -380,7 +381,67 @@ func isRevisionCountSame(fromClient, toClient *replicaClient.ReplicaClient, disk
...
+	if err != nil {
+		return err
+	}
+
+	snapshots := map[string]*types.Snapshot{}
+	var rwSize, woSize int64
+
+	for i := range disks {
+		disk := disks[len(disks)-1-i]
+		if _, ok := rwReplica.Disks[disk]; ok {
```
Comment (vishnuitta) [inline sync/sync.go:399]: lets handle !ok by returning err even though it shouldn't happen

File: sync/sync.go [hunk 15]
```diff
@@ -380,7 +381,67 @@ func isRevisionCountSame(fromClient, toClient *replicaClient.ReplicaClient, disk
 	return true, nil
 }
 
+func (t *Task) initalizeSyncProgress(fromClient, toClient *replicaClient.ReplicaClient, disks []string) error {
```
Comment (vishnuitta) [inline sync/sync.go:384]: add function comments

File: types/types.go [hunk 16]
```diff
@@ -125,6 +129,20 @@ type DiskInfo struct {
 	RevisionCounter int64    `json:"revisionCount"`
 }
 
+type Snapshot struct {
```
Comment (vishnuitta) [inline types/types.go:132]: add comments for new structures here

File: sync/sync.go [hunk 17]
```diff
@@ -380,7 +381,67 @@ func isRevisionCountSame(fromClient, toClient *replicaClient.ReplicaClient, disk
...
+	}
+
+	syncInfo := types.SyncInfo{
+		RWReplica:            fromClient.GetAddress(),
+		WOReplica:            toClient.GetAddress(),
+		Snapshots:            snapshots,
+		RWSnapshotsTotalSize: strconv.FormatInt(rwSize, 10),
+		WOSnapshotsTotalSize: strconv.FormatInt(woSize, 10),
+	}
+	rebuild.AddSyncInfo(&syncInfo)
```
Comment (vishnuitta) [inline sync/sync.go:435]: lets add a comment that this memory is not cleared to avoid unnecessary locking code due to its access in rebuild.GetReplicaInfo

**Pull request merged**